pilot/pkg/bootstrap/istio_ca.go

```diff
@@ -286,6 +289,181 @@ func (s *Server) loadRemoteCACerts(caOpts *caOptions, dir string) error {
+		return create
+	} else if event.Op&fsnotify.Write == fsnotify.Write {
+		return write
+	} else if event.Op&fsnotify.Remove == fsnotify.Remove {
+		return remove
+	}
+
+	return invalid
+}
+
+// updateCACertsMap updates the map with cacerts files with
+// particular events only.
+func updateCACertsMap(s *Server, event fsnotify.Event) bool {
+	file, ok := isValidCACertsFile(event.Name)
+	if !ok {
+		return false
+	}
+
+	op := isValidCACertsEvent(event)
+	if op == invalid {
+		return false
+	}
+
+	s.cacertsMutex.Lock()
+	s.cacertsMap[file] = op
+	s.cacertsMutex.Unlock()
+	return true
+}
+
+func areCACertsFilesCreated(s *Server) bool {
+	if s.cacertsMap["ca-cert.pem"] == create &&
+		s.cacertsMap["ca-key.pem"] == create &&
+		s.cacertsMap["root-cert.pem"] == create &&
+		s.cacertsMap["cert-chain.pem"] == create {
+		return true
+	}
+
+	return false
+}
+
+func areCACertsFilesModified(s *Server) bool {
+	if s.cacertsMap["ca-cert.pem"] == write &&
+		s.cacertsMap["ca-key.pem"] == write &&
+		s.cacertsMap["root-cert.pem"] == write &&
+		s.cacertsMap["cert-chain.pem"] == write {
+		return true
+	}
+
+	return false
+}
+
+// handleEvent handles the events on cacerts related files.
+// If create/write(modified) event occurs, then it verifies that
+// newly introduced cacerts are intermediate CA which is generated
+// from cuurent root-cert.pem. Then it updates and keycertbundle
+// and generates new dns certs.
+// TODO(rveerama1): Add support for new ROOT-CA rotation also.
+func handleEvent(s *Server, event fsnotify.Event) {
+	ok := updateCACertsMap(s, event)
+	if !ok {
+		return
+	}
+
+	if areCACertsFilesCreated(s) || areCACertsFilesModified(s) {
+		log.Info("Update Istiod cacerts")
+
+		currentCABundle := s.CA.GetCAKeyCertBundle().GetRootCertPem()
+		newCABundle, err := ioutil.ReadFile(path.Join(LocalCertDir.Get(), "root-cert.pem"))
+		if err != nil {
+			log.Error("failed reading root-cert.pem: ", err)
+			return
+		}
+
+		// Only updating intermediate CA is supported now
+		if !bytes.Equal(currentCABundle, newCABundle) {
+			log.Info("Updating new ROOT-CA not supported")
+			return
+		}
+
+		err = s.CA.GetCAKeyCertBundle().UpdateNewPluggedInCACerts(
+			path.Join(LocalCertDir.Get(), "/ca-cert.pem"),
+			path.Join(LocalCertDir.Get(), "ca-key.pem"),
+			path.Join(LocalCertDir.Get(), "cert-chain.pem"),
+			path.Join(LocalCertDir.Get(), "root-cert.pem"))
+		if err != nil {
+			log.Error("Failed to update new Plug-in CA certs: ", err)
+			return
+		}
+
+		err = s.updatePluggedinRootCertAndGenKeyCert()
+		if err != nil {
+			log.Error("Failed generating plugged-in istiod key cert: ", err)
+			return
+		}
+	}
+}
+
+// handleCACertsFileWatch handles the events on cacerts files
+func (s *Server) handleCACertsFileWatch() {
+	s.cacertsMap = map[string]uint8{
+		"ca-cert.pem":    invalid,
+		"ca-key.pem":     invalid,
+		"root-cert.pem":  invalid,
+		"cert-chain.pem": invalid,
+	}
+
+	for {
+		select {
+		case event, ok := <-s.cacertsWatcher.Events:
```

Comment:
would it be good to debounce updates? i.e wait for some time after we receive the first event, to absorb all additional events before we update the map. What is fsnotify is noisy and generates a ton of events?

Reply:
A timer to wait until to get all desired state in this case is not required I think. As per https://pkg.go.dev/github.com/fsnotify/fsnotify there are five different states. Here we are interested in only few events, rest we will ignore.

Reply:
@rveerama1 ptal pkg/kube/inject/watcher.go 

Add a debounce and also you donot need to check all files create or write.

Reply:
@hzxuzhonghu Thank you, I will take a look and update it.

Reply:
@shankgan @hzxuzhonghu sorry for misunderstanding. I added debounce now. Reworked it. Please take a look.

Reply:
is there any way to add a unit test for this? This is the most complex function by far, as it covers - (a) Validation of certFile, (b) Updation of keyCertbundle (c) calls the event handler.

Reply:
I thought off adding while doing new root-ca rotation support. As I have to add two intermediat-ca and different root-ca set of test certs.

Reply:
I would strongly recommend adding the test now :). There might be delay between the two features and we don't want to have this kind of code untested.

Reply:
ok, I will do that in next PR before multi-root rotation. Thanks.